### DISCUSSION TITLE
Memory-hard key derivation should be on the client.

### DISCUSSION THREAD

**Discussion by nuts-n-bits** [pocketbase/pocketbase#2134] [Q&A]
So according to a discussion I had recently it seems pb is doing bcrypt on server side. Apparently this is gonna be a bottleneck for any mid-sized project with lots of auth activities -- after all bcrypt's raison d'etre is to be precisely a massive bottleneck. This is enough of a problem that pb is planning to halve the hardness for bcrypt, and it's clear to me that it's only gonna get worse, and if we don't do something then the hardness will have to be halved again at some point. 

Solution is pretty easy conceptually, just let the client do as expensive of a bcrypt key derivation as they wish, then upload the result, and then apply an hmac hash with some salt on the server. 

This will: 1) reduce server CPU time and especially memory load since bcrypt by design eats memory; 2) we can now use a  higher hardness factor, improving security (13 barely causes a hiccup on an old iPhone); 3) eliminate an easy DoS target.

Just by the way allow me to mention that if you search about this topic, many bad takes are out there, such as [basically all the answers on this page](https://security.stackexchange.com/questions/8596/https-security-should-password-be-hashed-server-side-or-client-side). I was a crypto researcher and the scheme I suggested passes my own squint test, but I'll be the first to admit that crypto is hard to get right, and sometimes I do have blind spots, so the implementation details have to be ironed out carefully, and examined by multiple pairs of eyeballs. But at the end of the day, KDF needs to be on the client side is just one of those things that has to happen sooner or later. 

p.s. there are some good takes out there, too. [1](https://stackoverflow.com/questions/53366437/clientside-password-hashing)

**Comment by ganigeorgiev** [comment]
I disagree with this approach. It complicates a lot of things with very little to none benefit.

Please provide example implementation flow and why do you think hashing on the client-side would be more secure.

**Reply by nuts-n-bits** [reply]
It's not more secure because its logic is any better. It's more secure because it allows you to use a higher hardness factor because your server is no longer clogged by the MHKDF.

This is basically all the essential logic.  
```
// client
let derived_key = bcrypt(raw_passwd)
authenticate("https://server.com/auth", username, derived_key)

// server
let (username, derived_key) = // from HTTP request
let derived_key_salted = hmac_sha256(msg = derived_key, key = salt)
fetch_from_db(username, derived_key_salted)
```
Not complicated by my standards.

**Reply by ganigeorgiev** [reply]
Are you suggesting that we should perform a user authentication by just matching the submitted hash against the database record?

**Reply by nuts-n-bits** [reply]
Please look at the `//server` code block. The database stores the output of an hmac hash function, which is irreversible. There is no pass-the-hash, (due to hmac), obviously no dictionary (due to bcrypt), and no rainbow table (due to salting).

**Reply by ganigeorgiev** [reply]
>  The database stores the output of an hmac hash function

But it will be a fixed string, right?

It doesn't matter with what and how many times you hash the password. You are practically treating the hash as plain password.

And again, we have to implement something in addition to the above to prevent timing attacks because we'll end up comparing the result hash 1:1 with the db entry.

**Reply by ganigeorgiev** [reply]
Additionally, every time the user generate a bcrypt hash it will be different since the cipher text and the salt are encoded as one string.

**Reply by nuts-n-bits** [reply]
If the question is whether the hash is deterministic, then yes. It has to be.
If you are worrying about "can an attacker log in if the auth db leaks" then no. the `hmac` will prevent it. 

Note that the key param of the hmac function does not imply that you can "decrypt" its output to obtain the bcrypt hash. hmac is irreversible even if you know the key.

**Reply by nuts-n-bits** [reply]
> Additionally, every time the user generate a bcrypt hash it will be different since the cipher text and the salt are encoded as one string.

In my design, the bcrypt salt is irrelevant. We can just do `bcrypt(hardness = 13, salt = secure_hash(password), password = password)`. Using a constant public salt is probably gonna work. In this case bcrypt only provides hardness, the hmac on the server provides the salting.

**Reply by ganigeorgiev** [reply]
> the hmac will prevent it.

It won't prevent this because the key will be stored on the server.

---

I'm sorry, but I think we are going in circle from the previous discussion in https://github.com/pocketbase/pocketbase/discussions/2105#discussioncomment-5373984.

We are not only decreasing the security but also complicating both the client-side and server-side integrations. For what? For slightly better password_verify performance? If that's what we want then we can just use a lower bcrypt round factor.

And again, this actually may not be so easy as you think because as mentioned previously most bcrypt implementations don't work with user provided salt and users will generate a new bcrypt hash everytime they want to login but that hash will be different then what we hashed with the hmac and have in the database.

**Reply by nuts-n-bits** [reply]
While not trying to discourage you from this discussion, may I ask if you've received professional or academic training on cryptography? 

I ask because the question you are posing (It is NOT OK to be deterministic) cannot be processed by me without your help.

**Reply by nuts-n-bits** [reply]
I care enough about this project that I might just attempt a robustness proof on my design, and if you are indeed a cryptographer then your scrutiny is extremely welcomed. (If not it's still welcomed)

**Reply by ganigeorgiev** [reply]
Feel free to open a PR and I'll take the time to review it. I'm not a cryptographer but as it is right now I just don't see how hashing on client-side would be better.

My takes for now against hashing on the client-side are:

- It lower the security (at the minimum the suggested approach as far as I understand is prone to timing attacks).
- Will have to ship a bcrypt or any other crypto library with the SDKs
- Will have to fetch somehow a public salt to pass to client-side bcrypt (and again, most bcrypt implementations I've used doesn't work with user provided salts)
- Those who don't use a SDK will have to generate somehow the hash on their own (sending requests with curl, Postman or any other rest client will become bothersome)

Hashing on the server-side doesn't come without costs, but it is easy to manage and there is no need for custom hashing implementation. It also allows easier future migration in case you decide to rewrite your app from PocketBase to something else.

**Reply by nuts-n-bits** [reply]
> Will have to ship a bcrypt or any other crypto library with the SDKs
> Those who don't use a SDK will have to generate somehow the hash on their own (sending requests with curl, Postman or any other rest client will become bothersome)

Both valid points. If we in the end don't adopt client side hashing, these will probably be why. But then again these libs are tiny. Being curl-unfriendly, though, will be a truly unfortunate side effect of this change.  

> It lower the security (at the minimum the suggested approach as far as I understand is prone to timing attacks).

Timing attack is out of the scope of our design, but mitigation is trivial and shouldn't be a problem. Not to mention we have to mitigate timing attacks even if we run KDF on the server.

> Will have to fetch somehow a public salt to pass to client-side bcrypt (and again, most bcrypt implementations I've used doesn't work with user provided salts)

A 'pseudosalt' deterministically derived from the password should do. I suspect a public static salt will do. So no need to fetch any salt, because salting is now the responsibility of the server. Note that this is not creating a custom bcrypt implementation. 

> It also allows easier future migration in case you decide to rewrite your app from PocketBase to something else.

Another true point. Indeed sending raw password will be the most compatible auth option, but then again both bcrypt and hmac are very standardized. It does add friction though.
